### DISCUSSION TITLE
Fake FabricLoader implementation (isolated from connector?)

### DISCUSSION THREAD

**Discussion by sleepingdragoninn** [Sinytra/ForgifiedFabricAPI#142] [Ideas]
I've got a niche use case and would personally benefit from a neo implementation of FabricLoader (config directory, mod metadata, that kind of thing) being taken from connector (where I assume it currently resides) and being implemented into a neo mod, similar to FFAPI. (in my head, this would then get JIJ'd and extended by connector but I'm not familiar with the architecture) 

Basically, I've got a neo 1.21 branch of my fabric 1.21 mod, and through FFAPI the diff between the two branches is astronomically low - this means that we can develop features on fabric, merge them for free over to the neo branch, and we have the guaranteed safety of the compiler and mcdev (and therefore, CI) to tell us if something is wrong, rather than full-connector's runtime translation.

The only parts of the *actual code* (read: not the fmj/toml and buildscript) that change are calls to FabricLoader, and I know there has to be a forgified implementation in connector itself, I just don't know how wired into the rest of connector it is.

It'd be nice to use this individually! If I could make an extra mod() class and set up the init of a fabric ModInitializer from it (so I don't have to touch those either), that'd be even better.

Interested to know the state!

**Comment by Su5eD** [comment]
I think I've got just what you're looking for! [Forgified Fabric Loader](https://github.com/Sinytra/ForgifiedFabricLoader) is a Neo implementation of (most) Fabric Loader APIs, including mod metadata (of both Fabric and Neo mods), environment properties (dev/prod status, config dir, game dir, dist and mappings).

It can run from inside JiJ as well, and is already [used](https://github.com/Sinytra/ForgifiedFabricAPI/blob/79a4c2b00b1cb2e8108219916e05bb02917f3b9f/build.gradle.kts#L130) like that in FFAPI. Connector uses shadowjar to include it so that it can be accessed from the SERVICE module layer, but handles disabling JiJ'd versions bundled by other mods to avoid conflicts. The only missing features afaik are entrypoints (which are called by Connector) and environment-specific code stripping.

If you're using FFAPI from maven, it should already be pulled in along with its other dependencies.

**Reply by sleepingdragoninn** [reply]
oh awesome! great to know, I was just looking in the wrong places - thanks!

I don't think I was getting `FabricLoader` earlier, so I'll have to see if that was some kind of issue on my side - maybe stale references from another branch?

**Reply by sleepingdragoninn** [reply]
Hmm, yeah, tested on my fork of placeholderapi - not sure why it's not being pulled in, it shows up in libraries and everything!

![image](https://github.com/user-attachments/assets/6bcccf33-bf3e-47e0-8a2c-b6c199d7707a)
![image](https://github.com/user-attachments/assets/c42c0b43-5a64-46b8-b21b-37c98cbbccfa)
![image](https://github.com/user-attachments/assets/3e2297f4-8bdc-4f96-809b-94bbe7e52d6f)

**Reply by Su5eD** [reply]
Can you paste the output of gradlew dependencies?

**Reply by Su5eD** [reply]
Looking at maven metadata I think it's missing from the compile scope on my side

**Reply by sleepingdragoninn** [reply]
here's [gradlew dependencies](https://gist.github.com/sisby-folk/72234be4bc2ca89cd002c21d96c5ae7a)

~~Out of curiosity around FFL, are the ModInitializers implemented in any way? or are they just dummies?~~ read things

**Reply by Su5eD** [reply]
Released FFAPI `0.100.7+2.0.8+1.21` that addresses the issue. Let me know if there's anything else that needs fixing

**Reply by sleepingdragoninn** [reply]
Thank you! I'll take a look

**Reply by sleepingdragoninn** [reply]
Works a charm! Here's the diffs for the mods [TextPlaceHolderAPI](https://github.com/sisby-folk/TextPlaceholderAPI/compare/1.21..1.21-neo) and [Glowcase](https://github.com/ModFest/glowcase/compare/1.21..1.21-neo) for your own curiosity - extremely clean, thank you. Closing!
